internal/oidc/keys.go

```diff
@@ -0,0 +1,204 @@
+package oidc
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+
+	"github.com/dgrijalva/jwt-go"
+	fositejwt "github.com/ory/fosite/token/jwt"
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/authelia/authelia/internal/configuration/schema"
+	"github.com/authelia/authelia/internal/utils"
+)
+
+// NewKeyManagerWithConfiguration when provided a schema.OpenIDConnectConfiguration creates a new KeyManager and adds an
+// initial key to the manager.
+func NewKeyManagerWithConfiguration(configuration *schema.OpenIDConnectConfiguration) (manager *KeyManager, err error) {
+	manager = NewKeyManager()
+
+	_, _, err = manager.AddActiveKeyData(configuration.IssuerPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+// NewKeyManager creates a new empty KeyManager.
+func NewKeyManager() (manager *KeyManager) {
+	manager = new(KeyManager)
+	manager.keys = map[string]*rsa.PrivateKey{}
+	manager.keySet = new(jose.JSONWebKeySet)
+
+	return manager
+}
+
+// KeyManager keeps track of all of the active/inactive rsa keys and provides them to services requiring them.
+// It additionally allows us to add keys for the purpose of key rotation in the future.
+type KeyManager struct {
```

Comment:
Please move the struct on top of the file.

Reply:
Done.